Use new song select (v2) for playlists

smoogipoo opened the pull request:
- [x] Depends on https://github.com/ppy/osu/pull/35313
- Resolves part of https://github.com/ppy/osu/issues/34035

This replaces `PlaylistsSongSelect` with `PlaylistsSongSelectV2` that uses the new song select implementation.

https://github.com/user-attachments/assets/5fbf174c-47ec-4bef-b0d4-720b2ceb1224

It's still a little bit rough around the edges:
- The playlist popup is a little bit awkward, I didn't want to dive into the details area right now because it's tied to an enum tab control.
- When the overlay footer buttons are pressed (mods/freemods), footer buttons to the left remain in place. This is a limitation of the footer that I tried to improve but ultimately failed at.
- The song select panels could use a `( + )` button to their left for quick addition to the playlist. @peppy could give guidance as to how to do this, as I can only think of adding more filter interfacing to `SongSelect`. Also didn't want to dive into this for the time being.

I've suffixed all new components with `V2`, and I plan to rename them after also replacing the multiplayer implementation.

smoogipoo committed:
Add initial screen + test structure

smoogipoo committed:
Add initial freemods button

smoogipoo committed:
Add initial freestyle button

smoogipoo committed:
Add overall mods/freemods/freestyle hookups

smoogipoo committed:
Add initial playlist button

smoogipoo committed:
Integrate v2 playlists song select with game

smoogipoo committed:
Fix tests

smoogipoo committed:
Fix padding as subscreen

smoogipoo committed:
Add screen title

smoogipoo committed:
Fix padding not considered in sizing

smoogipoo committed:
Adjust button styling

smoogipoo committed:
Fix playlist with >1 item getting cleared

smoogipoo committed:
Remove old implementation

smoogipoo committed:
Remove song select dependency from FooterButtonOptions

smoogipoo committed:
Isolate footer behaviour to ScreenStackFooter, support subscreens

smoogipoo committed:
Use new implementation in ScreenTestScene

smoogipoo committed:
Fix CI inspections

arflyte commented:
Huh weird, I thought I already dropped my thoughts on that. I believe using the new song select as is wouldn't work well; needs to add a permanent tray to list all the selected songs. And yeah, replacing the osu! logo with "add to playlist" would work well.

arflyte commented:
Yeah, hence I suggested adding a permanent tray somewhere to show all the selected songs. I imagine once the song is adding, it can show an animation of the song going from the carousel into the tray (or just pop up in a list).

smoogipoo committed:
Merge branch 'master' into new-playlists-song-select

smoogipoo committed:
Add simple "Add to playlist" button in the footer

smoogipoo committed:
Remove add button from playlist popup

smoogipoo committed:
Add simple tray

smoogipoo committed:
Merge branch 'master' into new-playlists-song-select

peppy committed:
Fix inspections

peppy committed:
Adjust tray to feel better

peppy committed:
Merge branch 'master' into new-playlists-song-select

peppy committed:
Always flash button when adding a new item to playlist

peppy committed:
Fix/remove tests broken by screen behaviour change

peppy committed:
Merge branch 'master' into new-playlists-song-select

peppy commented on osu.Game/Screens/OnlinePlay/FooterButtonFreestyleV2.cs:
```diff
@@ -0,0 +1,113 @@
+        private void load()
+        {
+            Text = "Freestyle";
+            Icon = FontAwesome.Solid.ExchangeAlt;
```
We probably need a different icon for this (or free mods).

smoogipoo committed:
Revert no longer necessary change

Additionally to fix the options button, I could either cache the
interface in PlaylistsSongSelectV2 or make the interface cache itself. I
went with the latter option.

smoogipoo committed:
Adjust styling of freestyle button

smoogipoo commented on osu.Game/Screens/OnlinePlay/FooterButtonFreestyleV2.cs:
```diff
@@ -0,0 +1,113 @@
+                            Origin = Anchor.Centre,
+                            Font = OsuFont.Torus.With(size: 14f, weight: FontWeight.Bold),
+                            Shear = -OsuGame.SHEAR,
+                        },
```
Adjusted the styling to be like the options button. Seems to work well.

peppy reviewed (commented):
As commented

smoogipoo committed:
Fix title overlap

Uses the same method of displaying the mod selection as the old song
select (as an overlay of the entire game).

smoogipoo committed:
Fix footer button overlap

The scheduled events don't fire because the content is hidden.

smoogipoo committed:
Hide mods wedge when empty

smoogipoo committed:
Hide add button when mod selects are shown

The important one is the required mod selection, whose overlay contents
overlap with the add button. The free mod selection would otherwise be
fine, if not for consistency.

peppy committed:
Fix tray showing first item too far to left

peppy committed:
Fix failing test case due to mod overlay being moved to game level

peppy committed:
Remove unused button

peppy committed:
Merge branch 'master' into new-playlists-song-select
